src/transformers/trainer.py

```diff
@@ -952,10 +965,23 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
         `create_scheduler`) in a subclass.
         """
         self.create_optimizer()
-        self.create_scheduler(
-            num_training_steps=num_training_steps,
-            optimizer=self.optimizer.optimizer if is_sagemaker_mp_enabled() and smp.state.cfg.fp16 else self.optimizer,
-        )
+        if is_sagemaker_mp_enabled():
+            if SMP_POST_1_10:
+                # If smp >= 1.10, then we unwrap the optimizer is fp16 is enabled
+                self.create_scheduler(
+                    num_training_steps=num_training_steps,
+                    optimizer=self.optimizer.optimizer if smp.state.cfg.fp16 else self.optimizer,
+                )
+            else:
+                self.create_scheduler(
+                    num_training_steps=num_training_steps,
+                    optimizer=self.optimizer,
+                )
+        else:
+            self.create_scheduler(
+                num_training_steps=num_training_steps,
+                optimizer=self.optimizer,
+            )
```

Comment:
This is way too complicated for just a change in the optimizer used:
```suggestion
        if IS_SAGEMAKER_MP_POST_1_10 and smp.state.cfg.fp16:
            optimizer = self.optimizer.optimizer
        else:
            optimizer = self.optimizer
        self.create_scheduler(num_training_steps=num_training_steps, optimizer=optimizer)
```

Reply:
Oh thank you! This is much nicer thanks!